src/CMakeLists.txt

```diff
@@ -95,14 +97,49 @@ set(BOX2D_HEADER_FILES
 	../include/box2d/b2_world_callbacks.h
 	../include/box2d/box2d.h)
 
-add_library(box2d STATIC ${BOX2D_SOURCE_FILES} ${BOX2D_HEADER_FILES})
-target_include_directories(box2d PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/../include)
-target_include_directories(box2d PRIVATE ${CMAKE_CURRENT_SOURCE_DIR})
+add_library(box2d ${BOX2D_SOURCE_FILES} ${BOX2D_HEADER_FILES})
+
+target_include_directories(box2d
+  PUBLIC
+    ${CMAKE_CURRENT_SOURCE_DIR}/../include
+    $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}>
+  PRIVATE
+    ${CMAKE_CURRENT_SOURCE_DIR}
+)
+
 set_target_properties(box2d PROPERTIES
 	CXX_STANDARD 11
     CXX_STANDARD_REQUIRED YES
     CXX_EXTENSIONS NO
+    VERSION ${PROJECT_VERSION}
+    SOVERSION ${PROJECT_VERSION_MAJOR}
+)
+
+generate_export_header(box2d
+  BASE_NAME b2
+  EXPORT_MACRO_NAME B2_API
+  EXPORT_FILE_NAME b2_api.h
+  STATIC_DEFINE B2_STATIC
```

Comment:
I tried it and it generated this file:
```
#ifndef B2_API_H
#define B2_API_H

#ifdef B2_STATIC
#  define B2_API
#  define B2_NO_EXPORT
#else
#  ifndef B2_API
#    ifdef box2d_EXPORTS
        /* We are building this library */
#      define B2_API __attribute__((visibility("default")))
#    else
        /* We are using this library */
#      define B2_API __attribute__((visibility("default")))
#    endif
#  endif

#  ifndef B2_NO_EXPORT
#    define B2_NO_EXPORT __attribute__((visibility("hidden")))
#  endif
#endif

#ifndef B2_DEPRECATED
#  define B2_DEPRECATED __attribute__ ((__deprecated__))
#endif

#ifndef B2_DEPRECATED_EXPORT
#  define B2_DEPRECATED_EXPORT B2_API B2_DEPRECATED
#endif

#ifndef B2_DEPRECATED_NO_EXPORT
#  define B2_DEPRECATED_NO_EXPORT B2_NO_EXPORT B2_DEPRECATED
#endif

#if 0 /* DEFINE_NO_DEPRECATED */
#  ifndef B2_NO_DEPRECATED
#    define B2_NO_DEPRECATED
#  endif
#endif

#endif /* B2_API_H */
```

I think this is a breaking change on Windows. Every project using static box2d lib will have to define `B2_STATIC`. CMake projects won't be affected since you added it to the public include paths. Everyone else will have to define it manually.

Also people importing box2d inside their source tree and not compiling with CMake will have to generate this file manually.

The solution to both problem would be to create the file yourself and add it to the repo. Instead of using `#if B2_STATIC` you could use `#ifndef B2_DYNAMIC`.

Reply:
If you use cmake, it's not a breaking change, the default behavior is still the same as before. If you want to use box2d, you have to compile it cleanly with cmake. If you want to do garbage when compiling cmake, you will have garbage at the end. That's a cheap price to pay to have many configurations available (shared/static in this case) for everybody.

The `b2_api.h` is generated and should be generated. It's a CMake feature and it's very useful! People wanting to compile box2d in a weird way (not with cmake) will have to create it themselves, it's not a box2d problem.

Reply:
If you *use* box2d in a project that doesn't use cmake (or even if you're using an old FindBox2d cmake module, [a lot of projects are shipping their own](https://www.google.com/search?q=findbox2d.cmake)) you now have to define B2_STATIC manually while it wasn't required before. That's a breaking change.

As for the header I think it's a shame to use code generation for such a simple macro. Most projects don't do that. Gcc [documentation](https://gcc.gnu.org/wiki/Visibility) provides an example.

Reply:
Well, the new version of Box2D changed the name of the headers, which is a bigger breaking change and already breaks all the existing `FindBox2D`. Time to move on and use the other patch.

Reply:
You're right. I think think it's a shame to rely on code generation for such a simple thing but it works :+1: